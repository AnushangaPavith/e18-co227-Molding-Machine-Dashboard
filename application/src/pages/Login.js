import React, { useRef, useState, useEffect } from 'react';
import useAuth from '../auth/UseAuth';
import { useNavigate, useLocation } from 'react-router-dom';
import axios from '../auth/axios';

const LOGIN_URL = '/api/users/login';

const Login = () => {

    const { setAuth } = useAuth();
    const navigate = useNavigate();
    const location = useLocation();

    const from = location.state?.from?.pathname || "/Home";

    const userRef = useRef();
    const errRef = useRef();
    
    const [user, setUser] = useState('');
    const [pwd, setPwd] = useState('');
    const [errMsg, setErrMsg] = useState('');

    useEffect(() => {
        userRef.current.focus();
    }, []);

    useEffect(() => {
        setErrMsg('');
    }, [user, pwd]);
    
    const handleSubmit = async (e) => {
        e.preventDefault();

        try {
            const response = await axios.post(LOGIN_URL, 
                JSON.stringify({empID: user, password: pwd}), 
                {
                    headers: { 'Content-Type' : 'application/json'},
                    // withCredentials: true
                }
            );
            console.log(JSON.stringify(response?.data));

            const accessToken = response?.data?.token;
            const admin = response?.data?.admin.data[0];

            console.log(admin);

            setAuth({ user, pwd, accessToken });

            setUser('');
            setPwd('');
            // setSuccess(true);
            navigate(from, { replace: true });

        } catch (error) {

            if(!error?.response) {
                setErrMsg('No Server Response.');
            } else if(error.response?.status === 400) {
                setErrMsg("Missing Data.");
            } else if(error.response?.status === 401) {
                setErrMsg("Unauthorized.");
            } else {
                setErrMsg("Login Failed.");
            }
            errRef.current.focus();
        }
    }

    return (
            <div className='login-container'>
            <h3>Login to your account</h3><br></br>

            <p ref={errRef} className={errMsg ? "errMsg" : "offScreen"} aria-live="assertive">{errMsg}</p>

            <form onSubmit={handleSubmit}>
                <input
                    className="input"
                    type= 'text'
                    id='username'
                    ref={userRef}
                    placeholder= 'Employee ID'
                    onChange={(e) => setUser(e.target.value)}
                    value = {user}
                    required
                />

                <input
                    className="input"
                    type= 'password'
                    id='password'
                    placeholder= 'Password'
                    onChange={(e) => setPwd(e.target.value)}
                    value = {pwd}
                    required
                />

                <button className='btn'>
                    Login
                </button> 
            </form>
        </div>
    )
}

export default Login;