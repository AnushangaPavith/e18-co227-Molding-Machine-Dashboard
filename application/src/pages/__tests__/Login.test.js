import { render, screen, cleanup } from "@testing-library/react";
import {BrowserRouter as Router} from 'react-router-dom';
import Login from "../Login";

afterEach(() => { cleanup(); })

test("render and count input text fields", async () => {
    render(
        <Router>
            <Login />
        </Router>);
    const inputFields = await screen.findAllByRole('textbox')
    expect(inputFields).toHaveLength(1);
})

test("render and count password fields", async () => {
    render(
        <Router>
            <Login />
        </Router>);
    const passwordField = await screen.findAllByRole('textbox', { type: 'password' })
    expect(passwordField).toHaveLength(1);
})

test("render and count buttons", async () => {
    render(
        <Router>
            <Login />
        </Router>);
    const buttonList = await screen.findAllByRole('button');
    expect(buttonList).toHaveLength(1);
})
