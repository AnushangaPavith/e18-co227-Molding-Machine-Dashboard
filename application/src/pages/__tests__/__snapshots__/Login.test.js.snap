// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render and take snapshot 1`] = `
<DocumentFragment>
  <div
    class="login-container"
  >
    <div
      class="MuiBox-root MuiBox-root-4 login-box"
    >
      <h3>
        Login to your account
      </h3>
      <br />
      <p
        aria-live="assertive"
        class="offScreen"
      />
      <form>
        <input
          class="input"
          id="username"
          placeholder="Employee ID"
          required=""
          type="text"
          value=""
        />
        <input
          class="input"
          id="password"
          placeholder="Password"
          required=""
          type="password"
          value=""
        />
        <div
          class="login-button-container"
        >
          <button
            class="btn"
            type="button"
          >
            Sign Up
          </button>
          <button
            class="btn"
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
